### DISCUSSION TITLE
Panel Properties - Some available, some not? - SOLVED

### DISCUSSION THREAD

**Discussion by spiffo-smith** [RomanKubiak/ctrlr#567] [Q&A] [answered]
This one has gone me perplexed, if I add some simple script to a button like below:

local myChannel = tostring(panel:getProperty("panelMidiOutputChannelDevice"))
console("MIDI Ch is "..myChannel)
local version = tostring(panel:getProperty("panelVersionMajor"))
console("Version Major is "..version)
local delay = tostring(panel:getProperty("panelMidiGlobalDelay"))
console("Global Delay is "..delay)
local thruH2H = tostring(panel:getProperty("panelMidiThruH2H"))
console("H2H is "..thruH2H)
local thruH2D = tostring(panel:getProperty("panelMidiThruH2D"))
console("H2D is "..thruH2D)
local thruD2H = tostring(panel:getProperty("panelMidiThruD2H"))
console("D2H is "..thruD2H)
local thruC2D = tostring(panel:getProperty("panelMidiThruC2D"))
console("C2D is "..thruC2D)
local thruD2D = tostring(panel:getProperty("panelMidiThruD2D"))
console("D2D is "..thruD2D)

I get values returned for the Output Channel, the Version Major and the Global Delay but all the MIDI Thru Properties just return nil

I'd like to be able to read and change certain MIDI Thru settings without overly confusing someone with the Main MIDI Settings Menu which imho is confusing. But it looks like some Panel Properties are available and others are not?

**Comment by dnaldoog** [comment]
Hi Spiffo,

Try this:

<pre>
local myChannel = L(panel:getPropertyString("panelMidiOutputChannelDevice"))
console("MIDI Ch is "..myChannel)
local version = L(panel:getPropertyString("panelVersionMajor"))
console("Version Major is "..version)
local delay = L(panel:getPropertyString("panelMidiGlobalDelay"))
console("Global Delay is "..delay)
local thruH2H = L(panel:getPropertyString("panelMidiThruH2H"))
console("H2H is "..thruH2H)
local thruH2D = L(panel:getPropertyString("panelMidiThruH2D"))
console("H2D is "..thruH2D)
local thruD2H = L(panel:getPropertyString("panelMidiThruD2H"))
console("D2H is "..thruD2H)
local thruC2D = L(panel:getPropertyString("panelMidiThruC2D"))
console("C2D is "..thruC2D)
local thruD2D = L(panel:getPropertyString("panelMidiThruD2D"))
console("D2D is "..thruD2D)
</pre>

**Reply by spiffo-smith** [reply]
Magic! so as in:

local sName = L(comp:getOwner():getName())

To get a Modulator's name.

So why do we need L(....) and sometimes we don't, I've looked around but cannot find the reason, what is the Capital L and brackets doing? Is that LUA or JUCE?

**Reply by dnaldoog** [reply]
If you run this code:

<pre>
console("type(1) = "..type(1))
console("type(\"hello world\") = "..type("hello world"))
console("type(true) = "..type(true))
console("type({}) = "..type({}))
console("type(mod) = "..type(mod))
console("type(panel:getPropertyString(\"panelMidiThruH2H\") = "..type(panel:getPropertyString("panelMidiThruH2H")))
</pre>

you get:

<pre>
type(1) = number
type("hello world") = string
type(true) = boolean
type({}) = table
type(mod) = userdata
type(panel:getPropertyString("panelMidiThruH2H") = userdata
</pre>

So the last two `type` are `userdata`.

To parse a **string** contained in `userdata` that an object method like `getPropertyString()` returns, you need to surround the returned string with `L()`  - not sure why though, but it must be a way of formulating the `userdata` to be acceptable for lua or Ctrlr!

**Reply by spiffo-smith** [reply]
So the L() is a Ctrlr specific thing?

**Reply by dnaldoog** [reply]
> So the L() is a Ctrlr specific thing?

Yes it must be, but I can't find it in the source code. Can you?

**Reply by damiensellier** [reply, marked as answer]
@dnaldoog  I saw that already. it's a class for LUAbind it refers to wrapForLua (L) 

https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/Lua/JuceClasses/LCore.cpp#L10


From the luaBind doc https://luabind.sourceforge.net/docs.html:
Everything that gets registered in lua is registered in a namespace (lua tables) or in the global scope (called module). All registrations must be surrounded by its scope. To define a module, the luabind::module class is used. It is used like this:

```
module(L)
[
	// declarations
];
```

**Reply by dnaldoog** [reply]
Thanks Damien!!

**Reply by dnaldoog** [reply]
Also this from Possemo https://ctrlr.org/forums/topic/whats-the-difference-here/#post-83092

**Reply by spiffo-smith** [reply]
OK, thanks everyone, that's as detailed as it needs to be for me, all a bit confusing, especially since sometimes you need it and sometimes you don't!
